### PR TITLE
Fix some bugs I found in the `ManagedSocket` implementation

### PR DESCRIPTION
**Pull request by nvie**
This PR fixes a couple of bugs in the `ManagedSocket` (aka "connection manager") implementation. This code is still unused, so this PR should be safe to merge as-is.

### CODE REVIEW

File: packages/liveblocks-core/src/connection.ts [hunk 1]
```diff
@@ -412,54 +418,103 @@ function createConnectionStateMachine<T extends BaseAuthResult>(
...
+      // Then, set up all the necessary event listeners, and wait for the
+      // "open" event to occur.
+      //
+      // When the "open" event happens, we're ready to transition to the
+      // OK state. This is done by resolving the Promise.
+      //
+      async (ctx) => {
+        let capturedPrematureEvent: IWebSocketEvent | null = null;
+
+        const promise = new Promise<[IWebSocketInstance, () => void]>(
```
Comment (jrowny) [inline packages/liveblocks-core/src/connection.ts:433]: I know that 90% of my critiques are about naming things, but can we call this something other than promise?

File: packages/liveblocks-core/src/connection.ts [hunk 2]
```diff
@@ -412,54 +418,103 @@ function createConnectionStateMachine<T extends BaseAuthResult>(
...
+                socket.removeEventListener("close", reject); // Remove (*)
+              };
+
+              // Resolve the promise, which will take us to the @ok.* state
+              resolve([socket, unsub]);
+            });
           }
+        );
+
+        return withTimeout(promise, SOCKET_CONNECT_TIMEOUT).then(
```
Comment (jrowny) [inline packages/liveblocks-core/src/connection.ts:494]: does this need a catch somewhere?

Comment (nvie) [reply packages/liveblocks-core/src/connection.ts:494]: It does not! The whole point here is that the promise _should_ fail in all cases but the happy path. When that happens, the machine will [take the failure transition](https://github.com/liveblocks/liveblocks/blob/208f9c0531b3e9e69e76d33ddfc19d414970ae48/packages/liveblocks-core/src/connection.ts#L529-L564) out of that state, and either retries-with-backoff, or (purposefully) stops retrying.

Comment (nvie) [reply packages/liveblocks-core/src/connection.ts:494]: That said, I kinda hate how long this `.onEnterAsync()` definition has become. Makes the structure of that call really hard to follow now. The structure is:

```
    .onEnterAsync(
      "@connecting.busy",
      tryConnectingTheSocket,  // a Promise
      ok => { /* transition like this */ },
      failure => { /* transition like this */ },
    )
```

Comment (jrowny) [reply packages/liveblocks-core/src/connection.ts:494]: in the failure case, what happens to the socket? Is it cleaned up? Are those reject listeners removed?

Comment (nvie) [reply packages/liveblocks-core/src/connection.ts:494]: Great question! If the promise succeeds, we retain a permanent reference to the socket, via [this assignment](https://github.com/liveblocks/liveblocks/blob/208f9c0531b3e9e69e76d33ddfc19d414970ae48/packages/liveblocks-core/src/connection.ts#LL524-L524). Otherwise, the socket instance will get garbage collected, so my understanding is that there is no need to remove the listeners.

File: packages/liveblocks-core/src/connection.ts (hunk 1 above)
Comment (nvie) [reply packages/liveblocks-core/src/connection.ts]: Fixed in e4060b6f8c63247cfbe3986e00511d6a74f85b62. I've renamed it to `connect$`, using the convention of naming promises (values that still need to be `await`'ed) with a `$`-suffix, which we also do elsewhere in code to make that more explicit.
